File: alice/gamma/sdk/python/gamma_sdk/inner/parser.py
# coding: utf-8

import re

import attr

from gamma_sdk.inner.lexer import lexer


class ParserSyntaxError(Exception):
    def __init__(self, cause):
        self.cause = cause
        super(Exception, self).__init__()


@attr.s(frozen=True)
class ParsedTree:
    expression = attr.ib()
    variables = attr.ib(factory=list)

    def to_regex(self):
        return r'^' + self.expression.to_regex() + r'$'


@attr.s(frozen=True)
class OrNode:
    alternatives = attr.ib(factory=list)

    def to_regex(self):
        return r'(?:{})'.format('|'.join(alternative.to_regex() for alternative in self.alternatives))


@attr.s(frozen=True)
class SequenceNode:
    sequence = attr.ib(factory=list)

    def to_regex(self):
        return r'(?:\b| )'.join(element.to_regex() for element in self.sequence)


@attr.s(frozen=True)
class MaybeNode:
    expression = attr.ib()

    def to_regex(self):
        if isinstance(self.expression, OrNode):
            return r'{}?'.format(self.expression.to_regex())
        return r'(?:{})?'.format(self.expression.to_regex())


@attr.s(frozen=True)
class AnyNode:
    def to_regex(self):
        return r'.*'


@attr.s(frozen=True)
class TextNode:
    value = attr.ib(type=str)
    inflect = attr.ib(type=bool, default=False)
    prefix = attr.ib(type=bool, default=False)
    suffix = attr.ib(type=bool, default=False)

    def to_regex(self):
        value = re.escape(self.value)
        if self.prefix:
            value = r'.*' + value
        if self.suffix:
            value = value + r'.*'
        return value


@attr.s(frozen=True)
class VariableNode:
    name = attr.ib(type=str)
    type = attr.ib(type=str)
    value = attr.ib(type=str, default=None)

    def to_regex(self):
        return r'(?P<{}>[^ ]+?)'.format(self.name)


# Recursive descent LL(1) parser
class Parser:

    def __init__(self, stream):
        self._stream = stream
        self._peek = None
        self.stop = False
        self.variables = []

    def peek(self):
        try:
            if self._peek is None:
                self._peek = next(self._stream)
            return self._peek
        except StopIteration:
            self.stop = True
            return None

    def move(self):
        try:
            if self._peek is None:
                next(self._stream)
            else:
                self._peek = None
        except StopIteration:
            self.stop = True
            return None

    # EXPRESSION := SUB_EXPRESSION | SUB_EXPRESSION <OR> EXPRESSION
    # SUB_EXPRESSION := ELEMENT | ELEMENT SUB_EXPRESSION
    # ELEMENT := <TEXT> | <VARIABLE> | <STAR> |
    #            <LPAREN> EXPRESSION <RPAREN> | <LBRACKET> EXPRESSION <RBRACKET>

    def parse(self):
        tree = ParsedTree(expression=self.expression(), variables=self.variables)
        if not self.stop:
            raise SyntaxError("Unknown symbol %s" % self.peek())
        return tree

    def expression(self):
        result = []
        while not self.stop:
            sub_expression = self.sub_expression()
            result.append(sub_expression)
            token = self.peek()
            if self.stop or token.type != 'OR':
                break
            self.move()

        if len(result) == 1:
            return result[0]
        return OrNode(alternatives=result)

    def sub_expression(self):
        result = []
        while not self.stop:
            element = self.element()
            if element is None:
                break
            result.append(element)
        if len(result) == 1:
            return result[0]
        return SequenceNode(sequence=result)

    def element(self):
        token = self.peek()
        if self.stop:
            return None
        if token.type == 'TEXT':
            self.move()
            return TextNode(
                value=token.value.value,
                inflect=token.value.inflect,
                prefix=token.value.prefix,
                suffix=token.value.suffix,
            )
        if token.type == 'STAR':
            self.move()
            return AnyNode()
        if token.type == 'VARIABLE':
            self.move()
            variable = VariableNode(name=token.value.name, type=token.value.type, value=token.value.value)
            self.variables.append(variable)
            return variable
        if token.type == 'LPAREN':
            self.move()
            expression = self.expression()
            if self.peek().type != 'RPAREN':
                raise ParserSyntaxError('No closing parenthesis')
            self.move()
            return expression
        if token.type == 'LBRACKET':
            self.move()
            expression = self.expression()
            if self.peek().type != 'RBRACKET':
                raise ParserSyntaxError('No closing bracket')
            self.move()
            return MaybeNode(expression=expression)
        return None


def parse(string):
    return Parser(lexer(string)).parse()

File: alice/gamma/server/sdk/matcher/syntax/parser.go
package syntax

import "golang.org/x/xerrors"

type ParseTree struct {
	Expression *Node
}

type NodeType uint8

const (
	NodeNop NodeType = iota
	NodeAlt
	NodeSeq
	NodeMaybe
	NodeAny
	NodeText
	NodeVar
)

type Node struct {
	Type     NodeType
	Children []*Node
	Child    *Node
	Value    string
	Flags    Flags
	Name     string
	VarType  string
}

func nop() *Node {
	return &Node{Type: NodeNop}
}

func alteration(children []*Node) *Node {
	return &Node{Type: NodeAlt, Children: children}
}

func sequence(children []*Node) *Node {
	return &Node{Type: NodeSeq, Children: children}
}

func maybe(child *Node) *Node {
	return &Node{Type: NodeMaybe, Child: child}
}

func any() *Node {
	return &Node{Type: NodeAny}
}

func text(value string, flags Flags) *Node {
	return &Node{Type: NodeText, Value: value, Flags: flags}
}

func variable(name string, type_ string, value string) *Node {
	return &Node{Type: NodeVar, Name: name, VarType: type_, Value: value}
}

// EXPRESSION := SUB_EXPRESSION | SUB_EXPRESSION <OR> EXPRESSION
// SUB_EXPRESSION := ELEMENT | ELEMENT SUB_EXPRESSION
// ELEMENT := <TEXT> | <VARIABLE> | <STAR> |
//            <LPAREN> EXPRESSION <RPAREN> | <LBRACKET> EXPRESSION <RBRACKET>

type parser struct {
	stream <-chan *Token
	peek_  *Token
	stop   bool
}

func (parser_ *parser) peek() *Token {
	ok := true
	if parser_.peek_ == nil {
		parser_.peek_, ok = <-parser_.stream
	}
	if !ok {
		parser_.stop = true
	}
	return parser_.peek_
}

func (parser_ *parser) move() {
	ok := true
	if parser_.peek_ == nil {
		_, ok = <-parser_.stream
	} else {
		parser_.peek_ = nil
	}
	if !ok {
		parser_.stop = true
	}
}

const defaultCapacity = 8

func (parser_ *parser) expression() (*Node, error) {
	result := make([]*Node, 0, defaultCapacity)
	for !parser_.stop {
		subExpression, err := parser_.subExpression()
		if err != nil {
			return nil, err
		}
		if subExpression != nil && subExpression.Type != NodeNop {
			result = append(result, subExpression)
		}
		token := parser_.peek()
		if parser_.stop || token == nil || token.Type != TokenOr {
			break
		}
		parser_.move()
	}
	if len(result) == 0 {
		return nop(), nil
	}
	if len(result) == 1 {
		return result[0], nil
	}
	return alteration(result), nil
}

func (parser_ *parser) subExpression() (*Node, error) {
	result := make([]*Node, 0, defaultCapacity)
	for !parser_.stop {
		element, err := parser_.element()
		if err != nil {
			return nil, err
		}
		if element == nil {
			break
		}
		if element.Type != NodeNop {
			result = append(result, element)
		}
	}
	if len(result) == 0 {
		return nop(), nil
	}
	if len(result) == 1 {
		return result[0], nil
	}
	return sequence(result), nil
}

func (parser_ *parser) element() (*Node, error) {
	token := parser_.peek()
	if parser_.stop {
		return nil, nil
	}
	if token == nil {
		return nil, nil
	}
	switch token.Type {
	case TokenText:
		parser_.move()
		return text(token.Value, token.Flags), nil
	case TokenVariable:
		parser_.move()
		return variable(token.Name, token.VarType, token.Value), nil
	case TokenStar:
		parser_.move()
		return any(), nil
	case TokenLParen:
		parser_.move()
		expression, err := parser_.expression()
		if err != nil {
			return nil, err
		}
		nextToken := parser_.peek()
		if nextToken == nil || nextToken.Type != TokenRParen {
			return nil, xerrors.Errorf("syntax error: no closing parenthesis at %d", token.End)
		}
		parser_.move()
		return expression, nil
	case TokenLBracket:
		parser_.move()
		expression, err := parser_.expression()
		if err != nil {
			return nil, err
		}
		nextToken := parser_.peek()
		if nextToken == nil || nextToken.Type != TokenRBracket {
			return nil, xerrors.Errorf("syntax error: no closing bracket at %d", token.End)
		}
		parser_.move()
		if expression == nil || expression.Type == NodeNop {
			return expression, nil
		}
		return maybe(expression), nil
	case TokenError:
		return nil, xerrors.Errorf("unknown symbol: %v at %d", token.Value, token.End)
	}
	return nil, nil
}

func Parse(input string) (*ParseTree, error) {
	parser := parser{
		peek_:  nil,
		stop:   false,
		stream: lexer(input),
	}
	expression, err := parser.expression()
	if err != nil {
		return nil, err
	}
	if !parser.stop {
		return nil, xerrors.Errorf("syntax error: %v", parser.peek())
	}
	return &ParseTree{expression}, nil
}
